server/shared/src/test/scala/org/http4s/server/middleware/CSRFSuite.scala

```diff
@@ -351,24 +366,42 @@ class CSRFSuite extends Http4sSuite {
     }
 
     test("pass through and embed a slightly different token for a safe request") {
-      for {
+      val program = for {
         csrfCatchFailure <- csrfCatchFailureIO
         csrf <- csrfIO
         oldToken <- csrfCatchFailure.generateToken[IO]
         oldRaw <- csrfCatchFailure.extractRaw[IO](unlift(oldToken)).flatMap(IO.fromEither)
-        response <- csrfCatchFailure.validate()(dummyRoutes)(
-          csrf.embedInRequestCookie(passThroughRequest, oldToken)
-        )
-        newCookie =
-          response.cookies
-            .find(_.name == cookieName)
-            .getOrElse(ResponseCookie("invalid", "Invalid2"))
-        newToken = newCookie.content
-        newRaw <- csrfCatchFailure.extractRaw[IO](newToken).flatMap(IO.fromEither)
-      } yield {
-        assertEquals(response.status, Status.Ok)
-        assertNotEquals(oldToken.toString, newToken)
-        assertEquals(oldRaw, newRaw)
+      } yield (csrfCatchFailure, csrf, oldToken, oldRaw)
+
+      TestControl.execute(program).flatMap { control =>
+        for {
+          _ <- control.advanceAndTick(1.millis)
```

Comment:
Btw, this is a bit of an anti-pattern. Can we write this test case using `executeEmbed(...)` and `IO.sleep(...)` instead of `advanceAndTick`?

Reply:
Oh, really? Wasn't aware that this usage of `TestControl` is considered an anti-pattern 😓 Is this piece of knowledge presented on some CE resources?

Reply:
Yes, see here. Unless this is an advanced situation, `executeEmbed` is the preferred way.

https://typelevel.org/cats-effect/docs/core/test-runtime#full-execution

> The first sort of test we will attempt to write comes in the form of a complete execution. This is generally the most common scenario, in which most of the power of `TestControl` is unnecessary
...
> For more advanced cases, `executeEmbed` may not be enough to properly measure the functionality under test.

Reply:
`TestControl#execute` + `advanceAndTick` seems less tricky than `TestControl#executeEmbed` + `IO.sleep`. Or maybe I'm just biased about any usage of `IO.sleep` anywhere. But I really wonder in which ways this is precisely anti-pattern...

Reply:
> But I really wonder in which ways this is precisely anti-pattern...

You can ask Fabio about it more on the Discord. For example see comment here.
https://discord.com/channels/632277896739946517/632310980449402880/947533678546939935

Maybe "anti-pattern" was too strong a word. I meant it in the sense: using advanced, complex APIs when simple APIs will work is an anti-pattern.